### PR TITLE
partially move type outlives handling into trait solving

### PR DESCRIPTION
**Pull request by lcnr**
r? @nikomatsakis cc @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
```
 1,834,531  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection::traits::fulfill::FulfillProcessor, rustc_data_structures::obligation_forest::Outcome<rustc_trait_selection::traits::fulfill::PendingPredicateObligation, rustc_infer::traits::FulfillmentErrorCode>>
-1,237,662  ???:<rustc_infer::infer::combine::Generalizer as rustc_middle::ty::relate::TypeRelation>::relate_with_variance::<rustc_middle::ty::subst::GenericArg>
   960,255  ???:<rustc_middle::ty::subst::GenericArg as rustc_middle::ty::relate::Relate>::relate::<rustc_infer::infer::combine::Generalizer>
   880,869  ???:<rustc_infer::infer::lexical_region_resolve::LexicalResolver>::infer_variable_values
  -857,849  ???:rustc_infer::infer::lexical_region_resolve::resolve
   820,728  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::compress::<<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::process_obligations<rustc_trait_selection::traits::fulfill::FulfillProcessor, rustc_data_structures::obligation_forest::Outcome<rustc_trait_selection::traits::fulfill::PendingPredicateObligation, rustc_infer::traits::FulfillmentErrorCode>>::{closure
   662,607  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::apply_rewrites
  -564,995  ???:<rustc_infer::infer::equate::Equate as rustc_middle::ty::relate::TypeRelation>::relate_with_variance::<rustc_middle::ty::subst::GenericArg>
   551,065  ???:<rustc_middle::ty::subst::GenericArg as rustc_middle::ty::relate::Relate>::relate::<rustc_infer::infer::equate::Equate>
  -445,653  ???:<rustc_middle::ty::subst::GenericArg>::walk_shallow
   432,159  ???:<rustc_middle::ty::context::CtxtInterners>::intern_predicate
   404,411  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::uninlined_mark_dependents_as_waiting
   342,809  ???:<rustc_trait_selection::traits::fulfill::FulfillProcessor>::process_changed_obligations
  -327,517  ???:<rustc_infer::infer::combine::CombineFields>::instantiate
   272,601  ???:<rustc_infer::infer::equate::Equate as rustc_middle::ty::relate::TypeRelation>::tys
   264,798  ???:<rustc_infer::infer::InferCtxt>::nested_outlives_obligations
  -182,764  ???:rustc_infer::infer::outlives::components::compute_components_recursive
   164,839  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::find_cycles_from_node::<rustc_trait_selection::traits::fulfill::FulfillProcessor>
   131,944  ???:<rustc_infer::infer::InferCtxt>::resolve_regions_and_report_errors
   123,166  ???:free
  -112,131  ???:<rustc_middle::ty::context::TyCtxt>::intern_substs
  -105,280  ???:<alloc::vec::Vec<(rustc_middle::ty::sty::RegionVid, rustc_middle::ty::sty::RegionVid)>>::retain::<<rustc_infer::infer::lexical_region_resolve::LexicalResolver>::expansion::{closure
    99,428  ???:<hashbrown::raw::RawTable<(rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::Predicate>, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::Predicate>, rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::Predicate>, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
    95,105  ???:malloc
```

File: compiler/rustc_infer/src/infer/outlives/obligations.rs [hunk 1]
```diff
@@ -186,6 +191,146 @@ impl<'cx, 'tcx> InferCtxt<'cx, 'tcx> {
...
+        &self,
+        cause: &ObligationCause<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+        ty: Ty<'tcx>,
+        region: ty::Region<'tcx>,
+    ) -> Vec<PredicateObligation<'tcx>> {
+        let to_obligation = |pred: ty::Binder<'tcx, ty::PredicateKind<'tcx>>| {
+            Obligation::new(cause.clone(), param_env, pred.to_predicate(self.tcx))
+        };
+        let ty_outlives = |ty| {
```
Comment (oli-obk) [inline compiler/rustc_infer/src/infer/outlives/obligations.rs:205]: maybe just make this eagerly recurse here? That should save a bit of back-and-forth between this function and the trait solver.

**Comment by lcnr** [comment]
it's still slightly worse, though at a point where it's close to meaningless. One more improvement:

@bors try @rust-timer queue

**Comment by lcnr** [comment]
alright, good enough for me not to bother, time to improve the diagnostics

File: src/test/ui/regions/regions-assoc-type-static-bound-in-trait-not-met.stderr [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
-LL |     type Value: 'static;
-   |                 ^^^^^^^
+LL | impl<'a> Foo for &'a i32 {
+   |      ^^
+   = note: but lifetime parameter must outlive the static lifetime
 
 error: aborting due to previous error
 
-For more information about this error, try `rustc --explain E0477`.
+For more information about this error, try `rustc --explain E0478`.
```
Comment (yjhn) [inline src/test/ui/regions/regions-assoc-type-static-bound-in-trait-not-met.stderr:16]: I think this PR could also fix `E0478` to not show `must outlive the static lifetime` error message. See #41816.

**Comment by Dylan-DPC** [comment]
@lcnr any updates on this?

**Comment by lcnr** [comment]
not really, can close this for now

**Pull request closed without merging**